src/core/Entry.cpp

```diff
@@ -736,66 +834,81 @@ QString Entry::maskPasswordPlaceholders(const QString &str) const
+}
+
+QString Entry::resolveUrlPlaceholder(const QString &str, Entry::PlaceholderType placeholderType) const
+{
+    if (str.isEmpty())
+        return QString();
+
+    const QUrl qurl(str);
+    switch (placeholderType) {
+    case PlaceholderType::UrlWithoutScheme:
+        return qurl.toString(QUrl::RemoveScheme | QUrl::FullyDecoded);
+    case PlaceholderType::UrlScheme:
+        return qurl.scheme();
+    case PlaceholderType::UrlHost:
+        return qurl.host();
+    case PlaceholderType::UrlPort:
+        return QString::number(qurl.port());
+    case PlaceholderType::UrlPath:
+        return qurl.path();
+    case PlaceholderType::UrlQuery:
+        return qurl.query();
+    case PlaceholderType::UrlFragment:
+        return qurl.fragment();
+    case PlaceholderType::UrlUserInfo:
+        return qurl.userInfo();
+    case PlaceholderType::UrlUserName:
+        return qurl.userName();
+    case PlaceholderType::UrlPassword:
+        return qurl.password();
+    default:
+        Q_ASSERT(false);
+        break;
+    }
+
+    return QString();
+}
+
+Entry::PlaceholderType Entry::placeholderType(const QString &placeholder) const
+{
+    if (!placeholder.startsWith(QLatin1Char('{')) || !placeholder.endsWith(QLatin1Char('}'))) {
+        return PlaceholderType::NotPlaceholder;
+    } else if (placeholder.startsWith(QLatin1Literal("{S:"))) {
+        return PlaceholderType::CustomAttribute;
+    } else if (placeholder.startsWith(QLatin1Literal("{REF:"))) {
+        return PlaceholderType::Reference;
+    }
+
+    static const QMap<QString, PlaceholderType> placeholders {
+        { QStringLiteral("{TITLE}"), PlaceholderType::Title },
+        { QStringLiteral("{USERNAME}"), PlaceholderType::UserName },
+        { QStringLiteral("{PASSWORD}"), PlaceholderType::Password },
+        { QStringLiteral("{NOTES}"), PlaceholderType::Notes },
+        { QStringLiteral("{TOTP}"), PlaceholderType::Totp },
+        { QStringLiteral("{URL}"), PlaceholderType::Url },
+        { QStringLiteral("{URL:RMVSCM}"), PlaceholderType::UrlWithoutScheme },
+        { QStringLiteral("{URL:WITHOUTSCHEME}"), PlaceholderType::UrlWithoutScheme },
+        { QStringLiteral("{URL:SCM}"), PlaceholderType::UrlScheme },
+        { QStringLiteral("{URL:SCHEME}"), PlaceholderType::UrlScheme },
+        { QStringLiteral("{URL:HOST}"), PlaceholderType::UrlHost },
+        { QStringLiteral("{URL:PORT}"), PlaceholderType::UrlPort },
+        { QStringLiteral("{URL:PATH}"), PlaceholderType::UrlPath },
+        { QStringLiteral("{URL:QUERY}"), PlaceholderType::UrlQuery },
+        { QStringLiteral("{URL:FRAGMENT}"), PlaceholderType::UrlFragment },
+        { QStringLiteral("{URL:USERINFO}"), PlaceholderType::UrlUserInfo },
+        { QStringLiteral("{URL:USERNAME}"), PlaceholderType::UrlUserName },
+        { QStringLiteral("{URL:PASSWORD}"), PlaceholderType::UrlPassword }
+    };
+
+    PlaceholderType result = placeholders.value(placeholder.toUpper(), PlaceholderType::Unknown);
```

Comment:
Variable assignment is redundant. Directly return the result.

Reply:
Done